File: code/scenes/src/main/java/org/fbme/scenes/controllers/components/ComponentController.java
package org.fbme.scenes.controllers.components;

import jetbrains.mps.nodeEditor.cells.EditorCell;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.util.function.Function;

public interface ComponentController<FormT> {

    @NotNull
    Rectangle getBounds(FormT form);

    boolean canStartMoveAt(FormT form, int x, int y);

    @NotNull
    FormT translateForm(FormT originalForm, int dx, int dy);

    Function<Point, FormT> transformFormAt(FormT originalForm, int x, int y);

    @NotNull
    EditorCell getComponentCell();

    void updateCellWithForm(FormT form);

    void updateCellSelection(boolean selected);

    void paintTrace(Graphics g, FormT form);
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/components/ComponentController.kt
package org.fbme.scenes.controllers.components

import jetbrains.mps.nodeEditor.cells.EditorCell
import java.awt.Graphics
import java.awt.Point
import java.awt.Rectangle
import java.util.function.Function

interface ComponentController<FormT> {
    fun getBounds(form: FormT): Rectangle
    fun canStartMoveAt(form: FormT, x: Int, y: Int): Boolean
    fun translateForm(originalForm: FormT, dx: Int, dy: Int): FormT
    fun transformFormAt(originalForm: FormT, x: Int, y: Int): Function<Point, FormT>
    val componentCell: EditorCell
    fun updateCellWithForm(form: FormT)
    fun updateCellSelection(selected: Boolean)
    fun paintTrace(g: Graphics?, form: FormT)
}
